src/main/kotlin/net/maizegenetics/phgv2/api/HaplotypeGraph.kt

```diff
@@ -69,28 +73,47 @@ class HaplotypeGraph(hvcfFiles: List<String>) {
     fun ranges(): List<ReferenceRange> = refRangeMap.keys.sorted()
 
     /**
-     * Returns a hapId -> sample list map for the given ReferenceRange.
-     * Returned Map<hapId, List<sampleName>>
+     * Returns a map of contig to a sorted list of ReferenceRanges in each contig
+     */
+    fun rangesByContig(): Map<String, List<ReferenceRange>> {
+        return ranges().groupBy { refrange -> refrange.contig }.mapValues { (_, rangeList) -> rangeList.sorted() }
+    }
+
+    /**
+     * A sorted set of all SampleGametes in the graph
      */
-    fun hapIdToSamples(range: ReferenceRange, gameteId: Int = 0): Map<String, List<String>> {
+    fun sampleGametesInGraph(): SortedSet<SampleGamete> {
+        val gameteSet = sortedSetOf<SampleGamete>()
+        for (refrange in ranges()) {
+            gameteSet.addAll(hapIdToSampleGametes(refrange).values.flatten())
+        }
+        return gameteSet
+    }
+
+    fun hapIdToSampleGametes(range:ReferenceRange): Map<String, List<SampleGamete>> {
         val rangeId = refRangeMap[range]
         require(rangeId != null) { "hapIdToSamples: range: $range not found" }
-        val result = mutableMapOf<String, MutableList<String>>()
-        rangeByGameteIdToHapid[rangeId].forEachIndexed { sampleId, hapIdList ->
-            result.getOrPut(hapIdList[gameteId]) { mutableListOf() }.add(sampleNames[sampleId])
+
+        val result = mutableMapOf<String, MutableList<SampleGamete>>()
+        for (sampleId in rangeByGameteIdToHapid[rangeId].indices) {
+            for (gameteId in rangeByGameteIdToHapid[rangeId][sampleId].indices) {
+                val hapid = rangeByGameteIdToHapid[rangeId][sampleId][gameteId]
+                result.getOrPut(hapid) { mutableListOf() }.add(SampleGamete(sampleNames[sampleId], gameteId))
+            }
         }
         return result
     }
 
     /**
      * Returns the hapId for the sample in the specified ReferenceRange.
      */
-    fun sampleToHapId(range: ReferenceRange, sample: SampleGamete): String {
+    fun sampleToHapId(range: ReferenceRange, sample: SampleGamete): String? {
         val rangeId = refRangeMap[range]
         require(rangeId != null) { "sampleToHapId: range: $range not found" }
         val sampleId = sampleNameToIdMap[sample.name]
         require(sampleId != null) { "sampleToHapId: sample: $sample not found" }
-        return rangeByGameteIdToHapid[rangeId][sampleId][sample.gameteId]
+        val sampleHapids = rangeByGameteIdToHapid[rangeId][sampleId]
+        return if (sampleHapids.size > sample.gameteId) sampleHapids[sample.gameteId] else null
```

Comment:
Can you explain the logic here?  why "sampleHapids.size > sample.gameteId" ?  What is the expected value of sample.gameteid vs just checking that size is > 0?